### ISSUE TITLE
Implement a better description for 3rd party rules

### ISSUE THREAD

**Issue by EnTeQuAk** [mozilla/addons-linter#1183]
cc @muffinresearch @wagnerand and @jvillalobos on this for input.

In the past we implemented all rules for amo-validator and addons-linter ourselves so that we had total control over messages that were issued to the user.

With the most recent development that includes 3rd party rules (e.g https://github.com/mozilla/addons-linter/pull/1149 that implements very extensive eval-linting very easily) this power got lost a little.

Eslint doesn't provide any good mechanism to overwrite a rules message and since messages issued from an eslint rule don't have any exact ids we cannot filter the stream of messages ourselves.

My proposal: We do have the id of the rule that issued the message (a rule, say, no-innerHTML can issue multiple, different messages for various innerHTML scenarios but the general context usually stays the same) so let's use that and instead of replacing the complete message let's just enhance what we get from 3rd party rules by our own messages.

Practical example:

https://github.com/mozilla/addons-linter/blob/master/src/rules/javascript/eval-string-arg.js can and should be replaced by http://eslint.org/docs/rules/no-implied-eval

Currently issued message:
message: setTimeout or setInterval must have function as 1st arg
description: 'setTimeout and setInterval functions should be called only with function expressions as their first argument`

If implemented that'd be replaced by only "Implied eval. Consider passing a function instead of a string."

Now that is unfortunately a bad example… But let's say we do have a MDN page that describes this a bit better (like we usually do for API deprecations and stuff) what we could do is keep the eslint-message and supply our own ’description’ that is generic for all messages issued by that rule.

Another example, our current ’innerHTML’ related warnings are mostly something like "Unsafe assignment to innerHTML" but our amo-validator issues a bit more detailed description: "Due to both security and performance concerns, innerHTML may not be set using dynamic values which have                       not been adequately sanitized. This can lead to security issues or fairly serious performance                       degradation.

With my proposal we'd keep the message "Unsafe assignment to innerHTML" and supply our own description e.g from amo-validator that gives more detail on what we mean.

Any ideas, objections? I personally think that's the best option we have right now.

**Comment by EnTeQuAk** [comment]
Clarification: What I mean by "we cannot overwrite a message" is that we cannot overwrite specific messages, e.g we can't identify if a rule raised "Unsafe call to insertAdjacentHTML" or "Unsafe assignment to innerHTML" but we totally can set the same message for both cases.

```
16:24 <TheOne> ah, I *personally* think that should be fine
16:24 <TheOne> "I really expect different checks for the same rule to have the same message/description" (me, earlier)
```

If no one complains of has any objections I'll implement this for now so that we can overwrite `message` for all messages raised by a specific rule and apply a `description` too, or just one of both and merge it with what the 3rd party eslint rule issued. We can improve on this later anyway.

**Comment by EnTeQuAk** [comment]
I could not have phrased it better myself, that's correct.

Thanks a lot for the input!

**Comment by mstriemer** [comment]
I'm running into a somewhat similar problem in #1174 where I need to use a feature of eslint's reporter that is incompatible with our error codes. I want to do this:

```js
context.report(node, '{{api}} is unsupported', { api: 'gcm.register' });
```

Where `'gcm.register'` would be the name of an identifier we extracted from the source so it could be anything. Since we're looking up the code to get the message this doesn't work too well. Or at least I can't provide a code for it.

If we instead used the rule-id to get the description/message then that would let this work. The description would need to be something basic like "You used an API that isn't supported" (no interpolation) but at least the context would be there.

Maybe we don't need to update the old code to use this but it would probably be nice to have it be consistent. If this supported only adding to the message that would solve my problem.

```js
// messages/javascript.js
// I'm not sure how we want to store these but this comes to mind as an option.
const ruleUpdates = {
  'webextensions-unsupported-api': {
    code: 'UNSUPPORTED_API',
    description: 'This API is not implemented in Firefox and may '
                 + 'cause problems running your extension with Firefox.',
    legacyCode: null,
  },
};

// webextensions-unsupported-api.js
context.report(node, '{{api}} is unsupported', { api });

**Comment by EnTeQuAk** [comment]
I agree, something like that would be nice too. Very good point indeed, that's multiple things in one though. What you want should work with what I partially implemented in https://github.com/mozilla/addons-linter/pull/1189 but might need some more work to get it right since it's based off the need to overwrite eslint specific messages.

Theoretically you should be able to have a format-string in the `message` and `code` in a generic and static `UNSUPPORTED_API` and substitute them during run-time when you call `context.report` maybe? Another idea could be to generate those codes during run-time similar to how `messages/javascript:deprecatedAPI` does it?

**Comment by mstriemer** [comment]
> What you want should work with what I partially implemented in #1189

Looking through the changes it looks like it will work. I'll give it a try and get back to you.

> Theoretically you should be able to have a format-string in the message and code in a generic and static UNSUPPORTED_API and substitute them during run-time when you call context.report maybe?

This sounds like what eslint already does, but you can only set the message. From what I can see the `code` is an addons-linter thing, and it's really just the `message` but we process it more later. So I can only set either the `code` or the `message`.

> Another idea could be to generate those codes during run-time similar to how messages/javascript:deprecatedAPI does it?

The deprecated API ones work because all of the APIs are known beforehand. We could generate arbitrary codes at runtime temporarily but I think this is a better solution.

**Comment by EnTeQuAk** [comment]
This got merged with https://github.com/mozilla/addons-linter/pull/1189 - could use a few more improvements but that'll happen once we need it.
